Added a re-implementation of RIBES

alvations opened the pull request:
RIBES is a popular MT evaluation metric, especially for [Asian language pairs](http://lotus.kuee.kyoto-u.ac.jp/WAT). The official scripts can be found at http://www.kecl.ntt.co.jp/icl/lirg/ribes/

This re-implementation provides an educational documentation of the various parts of the algorithm described in the paper. 

Since the licenses differs for NLTK and RIBES, having RIBES code in verbatim would be technically wrong. Users are still encouraged to use the official script when evaluating their MT system.

Note that the emulation code was removed at https://github.com/alvations/nltk/commit/520a6760b3012aa0fd245264741ad4d9563645da at the suggestion of the original RIBES maintainers.

alvations committed:
Added RIBES score to the `translate` package

alvations committed:
Added RIBES to the translate.__init__.py

alvations committed:
Changes names

alvations committed:
Fixed typo in comments and added new line in docstring

alvations committed:
Added author information.

alvations committed:
Fixed typo in ribes formula, ** not *

alvations committed:
Fixed typo in docstring

alvations committed:
Added missing parenthesis in formula

alvations committed:
Bugfixes and better increasing sequence catcher

- Bugfixes in checking for larger context window in extracting the `worder` list
- Use ekhumoro's `igroups` catcher instead of the previous boundary hack.

alvations committed:
Update ribes_score.py

alvations committed:
Removed emulation code and update docstring

stevenbird commented:
@hoontw – would you mind reviewing this contribution please?

hoontw commented:
I will review this over the next few days.

hoontw commented on nltk/translate/ribes_score.py:
```diff
@@ -0,0 +1,311 @@
+    implementation when evaluating your machine translation system. Refer
+    to http://www.kecl.ntt.co.jp/icl/lirg/ribes/ for the official script.
+    
+    :param reference: a reference sentence
```
Should this be `references`?

hoontw commented on nltk/translate/ribes_score.py:
```diff
@@ -0,0 +1,311 @@
+    :type alpha: float
+    :param beta: hyperparameter used as a prior for the brevity penalty.
+    :type beta: float
+    :param character_based: a flag to emulate the original character-based RIBES
```
`character_based` doesn't seem to be relevant anymore

hoontw commented on nltk/translate/ribes_score.py:
```diff
@@ -0,0 +1,311 @@
+    :param beta: hyperparameter used as a prior for the brevity penalty.
+    :type beta: float
+    :param character_based: a flag to emulate the original character-based RIBES
+    :type character_based: boolean 
```
Would it help to have a doctest here to illustrate usage and the expected output?

hoontw commented on nltk/translate/ribes_score.py:
```diff
@@ -0,0 +1,311 @@
+        nkt = kendall_tau(worder)
+            
+        # Calculates the brevity penalty
+        bp = min(1.0, math.exp(1.0 - 1.0 * len(reference)/len(hypothesis)))
```
`bp = max(1.0, math.exp(1.0 - 1.0 * len(reference)/len(hypothesis)))` seems to make more sense to me

alvations replied:
Brevity penalty is to reduce the score by penalizing the long hypothesis that gets higher kendall's tau unigram precision. So smaller bp means lesser penaly and that means that the hypothesis is close to the reference in terms of length, thus `min(...)` here is right. `max(1.0, ...)` would always result in 1.0 because ` 0 < math.exp(1.0 - 1.0 * len(reference)/len(hypothesis)) < 1.0` is true for all length of hypothesis and reference.

alvations replied:
`min(...)` is right here because brevity penalty penalizes short hypothesis because precision is calculated as such  `precision = 1.0 * n / len(hyp)`. So the shorter the `len(hyp)` the higher the precision. These example will illustrate the penalty.

When hypothesis is close to the reference in length, we don't penalize the unigram precision as much, e.g.:

```
>>> import math
>>> lenref, lenhyp = 10, 9
>>> n = 5
>>> precision = 1.0 * n / lenhyp
>>> bp = min(1.0, math.exp(1.0 - 1.0 * lenref/lenhyp))
>>> precision
0.5555555555555556
>>> bp
0.8948393168143697
>>> bp * precision 
0.49713295378576094
```

But when it's much shorter, we should penalize it more:

```
>>> import math
>>> lenref, lenhyp = 10, 6
>>> n = 5
>>> precision = 1.0 * n / lenhyp
>>> precision
0.8333333333333334
>>> bp = min(1.0, math.exp(1.0 - 1.0 * lenref/lenhyp))
>>> bp
0.513417119032592
>>> bp * precision
0.4278475991938267
```

And if the hypothesis is super short, precision is can still be relatively high by it didn't match the longer reference so it needs to be penalize

```
>>> lenref, lenhyp = 10, 4
>>> n = 3
>>> precision = 1.0 * n / lenhyp
>>> precision
0.75
>>> bp = min(1.0, math.exp(1.0 - 1.0 * lenref/lenhyp))
>>> bp * precision 
0.16734762011132237
```

hoontw replied:
Sorry, you're right. I was thinking about the case when length of hypothesis > length of reference. My brain isn't working at this time of night :)

alvations replied:
No problem =) I just realized, I forgot this example that will show the rationale behind brevity penalty. Even when all ngrams in hypothesis matches, i.e. `n == 4; lenhyp == 4`, it is still quite a bad translation if it only translate a small bit of the sentence. 

```
>>> lenref, lenhyp = 10, 4
>>> n = 4
>>> precision = 1.0 * n / lenhyp
>>> bp = min(1.0, math.exp(1.0 - 1.0 * lenref/lenhyp))
>>> bp * precisio
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'precisio' is not defined
>>> bp * precision
0.22313016014842982
```

hoontw commented on nltk/translate/ribes_score.py:
```diff
@@ -0,0 +1,311 @@
+        if _ribes > _best_ribes: # Keeps the best score.
+            _best_ribes = _ribes
+        
+    return _best_ribes, p1, bp, worder, nkt
```
Some documentation about the return values would be nice :)

alvations replied:
Whoops that's my mistake, it should only return _best_ribes. The rest of the return values was used for debugging the code =)

hoontw commented on nltk/translate/ribes_score.py:
```diff
@@ -0,0 +1,311 @@
+    w.r.t. the list of reference words.
+    
+    Below is (H0, R0) example from the Isozaki et al. 2010 paper, 
+    note the examples are indexed from 1th but the results here starts from 0th:
```
grammar nit: `are indexed from 1 but the results here are indexed from 0:`

hoontw commented on nltk/translate/ribes_score.py:
```diff
@@ -0,0 +1,311 @@
+    # This is used for matching context window later in the algorithm.
+    ref_ngrams = []
+    hyp_ngrams = []
+    for n in range(1, len(reference)+1):
```
The original paper only considered unigrams and bigrams. Do current implementations consider ngrams of any length?

alvations replied:
The current implementation from the official maintainers considers ngrams of any length. 

The results reported in the paper considers all ngram lengths too, otherwise it will be hard to achieve the correlation results they've gotten. Only in the examples, they simplify the explanation and they use unigrams and bigrams.

hoontw replied:
I see, thanks!

hoontw commented on nltk/translate/ribes_score.py:
```diff
@@ -0,0 +1,311 @@
+        elif hypothesis.count(h_word) == reference.count(h_word) == 1:
+            worder.append(reference.index(h_word))
+        else:
+            # Note: range(1, max(i, hyp_len-i+1)) is the range of window sizes.
```
nit: could replace comment with a desciptive variable. Example:

```
max_window_size = max(i, hyp_len-i+1)
for window in range(1, max_window_size):
```

alvations committed:
Edited docstring

- Corrected reference -> references in docstring
- Removed the character_based parameter from docstring

alvations committed:
Added return type for the docstring

alvations committed:
Remove unnecessary returns from the ribes func.

alvations committed:
Removed doctest snippet

alvations committed:
Corrected grammar in docstring

alvations committed:
Use the max_window_size variable instead of repeating the max(...)

hoontw commented on nltk/translate/ribes_score.py:
```diff
@@ -0,0 +1,311 @@
+    return worder
+
+ 
+def choose(n, k):
```
Wasn't a nCr function introduced into NLTK recently?

alvations replied:
Yep, I should import this properly. It's the same function.

alvations committed:
Removed choose function, use from nltk.util instead

alvations committed:
Corrected grammar in docstring

hoontw commented on nltk/translate/ribes_score.py:
```diff
@@ -0,0 +1,311 @@
+    """
+    items = iter(worder)
+    a, b = None, next(items, None)
+    result = [b]
```
If `worder` is `[]`, is it ok for this function to return `[None]`?

alvations replied:
If `worder == []`, `find_increasing_sequence()` will not return `None` or `[None]` but an empty generator because it doesn't return a list but a generator. And the generator will be iterated later to calculate the kendall's tau, so when doing a `sum(iter([]))` it returns a 0. E.g.

```
>>> worder = []
>>> len(worder)
0
>>> def find_increasing_sequences(worder):
...     """
...     Given the *worder* list, this function groups monotonic +1 sequences. 
...     
...         >>> worder = [7, 8, 9, 10, 6, 0, 1, 2, 3, 4, 5]
...         >>> list(find_increasing_sequences(worder))
...         [(7, 8, 9, 10), (0, 1, 2, 3, 4, 5)]
...     
...     :param worder: The worder list output from word_rank_alignment
...     :param type: list(int)
...     """
...     items = iter(worder)
...     a, b = None, next(items, None)
...     result = [b]
...     while b is not None:
...         a, b = b, next(items, None)
...         if b is not None and a + 1 == b:
...             result.append(b)
...         else:
...             if len(result) > 1:
...                 yield tuple(result)
...             result = [b]
... 
>>> find_increasing_sequences([])
<generator object find_increasing_sequences at 0x7f9fbe71ca00>
>>> increasing_sequences = find_increasing_sequences([])
>>> sum(i for i in increasing_sequences)
0
>>> sum(iter([]))
0
```

hoontw replied:
Thanks for the detailed explanation and verifying it still works with edge cases. I'm rusty on Python generators.

alvations commented:
@hoontw thanks for looking through the code! Please continue to review if you have yet finished but I've got to go offline for a while. And I'll get back to the edits tomorrow.

hoontw commented on nltk/translate/ribes_score.py:
```diff
@@ -0,0 +1,280 @@
+        tau = 2 * num_increasing_pairs / num_possible pairs -1
+    
+    Note that the no. of increasing pairs can be discontinuous in the *worder*
+    list and each increase sequence can be tabulated as choose(len(seq), 2) no. 
```
grammar: each increasing sequence

hoontw commented on nltk/translate/ribes_score.py:
```diff
@@ -0,0 +1,280 @@
+    
+    Note that the no. of increasing pairs can be discontinuous in the *worder*
+    list and each increase sequence can be tabulated as choose(len(seq), 2) no. 
+    of increasing pair, e.g.
```
grammar: increasing pairs

hoontw commented on nltk/translate/ribes_score.py:
```diff
@@ -0,0 +1,280 @@
+        >>> round(kendall_tau(worder),3)
+        0.382
+    
+    To compare with the _ribes_kendall_tau():
```
I believe `_ribes_kendall_tau` has been removed

hoontw commented on nltk/translate/ribes_score.py:
```diff
@@ -0,0 +1,280 @@
+    Using the (H0,R0) and (H5, R5) example from the paper
+    
+        >>> worder =  [7, 8, 9, 10, 6, 0, 1, 2, 3, 4, 5]
+        >>> sum((wi - i)**2 for wi, i in zip(worder, range(len(worder))))
```
This line doesn't seem to belong in the doctest because it is about internal calculations. `worder` is also defined twice, so one of them could be removed.

alvations replied:
I've removed the extra `worder` initialization but the spearman line (sum((wi - i)**2 for wi, i in zip(worder, range(len(worder))))`) that outputs 350 comes from the example from the paper e.g. (H5, R5).

hoontw replied:
I think the extra worder initialization is still here.
For the sum of squares, I am not sure if users will be interested in the intermediate steps in deriving spearman's rho. I do not feel strongly about this, so if you want to leave it in, that's ok.

alvations committed:
Fix grammar in docstring

alvations committed:
Fix grammar in docstring

Added return type for `kendall_tau()`

alvations committed:
Remove sum d-square and extra worder in docstring

alvations committed:
Deleted extra comment on max_window_size

alvations commented:
@hoontw Thanks again for the reviews. I think I've fixed most of the suggested edits, I hope I didn't miss anything. Sorry that it took a while to revise because I was busy.
